File: PassedProblems/150-evaluate-reverse-polish-notation.cpp
/*
 * 150-evaluate-reverse-polish-notation.cpp
 *
 *  Created on: Jul 15, 2015
 *      Author: hanzhe
 */

#include <iostream>
#include <vector>
#include <stack>
#include <string>
#include <sstream>
using namespace std;

class Solution {
public:
    int evalRPN(vector<string>& tokens) {
    	stack<int> nums;
    	for(int i = 0 ; i < tokens.size(); i ++)
    	{
    		string s = tokens[i];
    		if(s == "+" || s == "-" || s == "*" || s == "/")
    		{
    			int o1 = nums.top();
    			nums.pop();
    			int o2 = nums.top();
    			nums.pop();
    			switch(s[0]){
    			case '+':
    				nums.push(o1 + o2);break;
    			case '-':
    				nums.push(o2 - o1);break;
    			case '*':
    				nums.push(o1 * o2);break;
    			case '/':
    				nums.push(o2 / o1);break;
    			}
    		}
    		else{
    			istringstream str(s);
    			int i;
    			str >> i;
    			nums.push(i);
    		}
    	}
    	return nums.top();
    }
};

File: leetcode/python/passedProblems/150-evaluate-reverse-polish-notation.py
#!/usr/bin/env python
# encoding: utf-8

"""
Evaluate the value of an arithmetic expression in Reverse Polish Notation.

Valid operators are +, -, *, /. Each operand may be an integer or another expression.

Some examples:
  ["2", "1", "+", "3", "*"] -> ((2 + 1) * 3) -> 9
  ["4", "13", "5", "/", "+"] -> (4 + (13 / 5)) -> 6
"""

# 网上答案,速度一般
class Solution(object):
    def evalRPN(self, tokens):
        stk = []
        opr = {'-': lambda x,y: x-y, '+': lambda x,y:x+y, '*': lambda x,y:x*y, '/': lambda x,y: x/y}
        for t in tokens:
            try:
                stk.append(float(t))
            except:
                stk.append(int(opr[t](stk.pop(-2), stk.pop(-1))))
        return int(stk[0])



# 时间效率很差，调了很多次，在/这里，系统要求 6/-132=0,13/5=2
class Solution(object):
    def evalRPN(self, tokens):
        """
        :type tokens: List[str]
        :rtype: int
        """
        ns = []
        for x in tokens:
            if x in "+-*/":
                tp2 = ns.pop()
                tp1 = ns.pop()
                if x == '+': ns.append(tp1+tp2)
                if x == '-': ns.append(tp1-tp2)
                if x == '*': ns.append(tp1*tp2)
                if x == '/': ns.append(int((1.0*tp1/tp2)))
                #if x == '/': ns.append((tp1/tp2))
                print x, tp1, tp2, ns[-1]
            else: ns.append(int(x))
        return ns[0]
